fabtests/unit/mr_cache_evict.c

```diff
@@ -0,0 +1,688 @@
+	 * reallocated and re-registered. The newly registered MR should not
+	 * have been cached.
+	 */
+	prev_buf = buf;
+	mem_free(buf);
+
+	buf = mem_alloc(type, mr_buf_size);
+	if (!buf) {
+		ret = -ENOMEM;
+		FT_UNIT_STRERR(err_buf, "mem_alloc failed", ret);
+		goto cleanup;
+	}
+
+	/* We NEED the same pointer to be returned for this test to be valid. */
+	if (buf != prev_buf) {
+		ret = -EFAULT;
+		FT_UNIT_STRERR(err_buf,
+			       "Failed to reallocate same virtual address",
+			       ret);
+		goto cleanup;
+	}
+
+	/* Verify reallocated MR registration time is close to the initial MR
+	 * registration time and greater than the cached MR registration time.
+	 */
+	ret = mr_register(buf, mr_buf_size, &realloc_mr, &realloc_mr_reg_time);
+	if (ret) {
+		FT_UNIT_STRERR(err_buf, "mr_register failed", ret);
+		goto cleanup;
+	}
+
+	FT_DEBUG("Reallocated MR registration time: %ld nsecs",
+		 realloc_mr_reg_time);
+
+	if (realloc_mr_reg_time <= CACHE_TIME_MAX_VALUE(mr_reg_time)) {
+		ret = -EEXIST;
+		FT_UNIT_STRERR(err_buf,
+			       "Reallocated MR registration time too low. "
+			       "Cached MR may have been incorrectly used.",
+			       ret);
+	} else {
+		testret = PASS;
+	}
+
+cleanup:
+	if (realloc_mr)
+		fi_close(&realloc_mr->fid);
+
+	if (cached_mr)
+		fi_close(&cached_mr->fid);
+
+	if (mr)
+		fi_close(&mr->fid);
+
+	if (buf)
+		mem_free(buf);
+
+	if (prime_mr)
+		fi_close(&prime_mr->fid);
+
+	if (prime_buf)
+		free(prime_buf);
+
+	return TEST_RET_VAL(ret, testret);
+}
+
+/* Run tests using MMAP, BRK, and SBRK. */
+static int mr_cache_mmap_test(void)
+{
+	return mr_cache_test(MMAP);
+}
+
+static int mr_cache_brk_test(void)
+{
+	return mr_cache_test(BRK);
+}
+
+static int mr_cache_sbrk_test(void)
+{
+	return mr_cache_test(SBRK);
+}
+
+struct test_entry test_array[] = {
+	TEST_ENTRY(mr_cache_mmap_test, "MR cache eviction test using MMAP"),
+	TEST_ENTRY(mr_cache_brk_test, "MR cache eviction test using BRK"),
+	TEST_ENTRY(mr_cache_sbrk_test, "MR cache eviction test using SBRK"),
+	{ NULL, "" }
+};
+
+static void usage(void)
+{
+	ft_unit_usage("mr_cache",
+		"Test a provider's ability to evict MR cache entries.\n"
+		"Evictions are verified using MMAP, BRK, and SBRK allocations.\n"
+		"Users should set the FI_MR_CACHE_MONITOR env variable to \n"
+		"change the memory monitor used by the MR cache.\n\n"
+		"With debug enabled, when running as root, the physical \n"
+		"address of the first page of the MMAP, BRK, and SBRK \n"
+		"allocation is returned. This can be used to verify the \n"
+		"underlying physical memory changes between MMAP, BRK, and \n"
+		"SBRK allocations. When running as non-root, the reported \n"
+		"physical address is always zero.");
+	FT_PRINT_OPTS_USAGE("-s <bytes>", "Memory region size to be tested.");
+}
+
+int main(int argc, char **argv)
+{
+	int ret;
+	int op;
+	int failed = 0;
+
+	/* On Linux, force malloc to avoid used sbrk/brk by setting
+	 * M_MMAP_THRESHOLD to 1. This will result in malloc using mmap. This
```

Comment:
nit: It's clearer to say that we force malloc to use mmap, rather than forcing it to avoid x and y.